File: 3° Anno/Basi Di Dati/Laboratorio/Michael_Posenato_Esami_Svolti/• 2019 4 Luglio/es3.py
# programma py che leggendo un prodotto da console, visualizza la citta
# risultato della query 2b (adattata al fatto che si fornisce un codice prodotto).

# se il prodotto non e' venduto o se e' venduto in piu' citta', il programma deve
# segnalare la cosa.

# se il prodotto non esiste, il programma deve segnalare l'errore e richiedere
# nuovamente il dato

import psycopg2
from myConfig import myHost, myDB, myUsr, myPsw

# connessione al db
with psycopg2.connect(host=myHost, database=myDB, user=myUsr, password=myPsw) as conn:
	with conn.cursor() as cur:

		# acquisizione id prodotto e relativo controllo
		idProdotto = ''
		checking = True

		while(checking):
			idProdotto = input('Inserisci ID prodotto: ')

			cur.execute('SELECT COUNT(*) FROM prodotto WHERE codice=%s', (idProdotto,))
			status = cur.fetchone()

			if status == None or status[0] != 1:
				print('l\'ID inserito non corrisponde a nessun prodotto. Ripetere inserimento')
				continue

			checking = False

		# prodotto valido: eseguo la query
		cur.execute(''' 
			SELECT 	DISTINCT n.citta
			FROM 	prodotto p 
					JOIN listino l ON l.prodotto = p.codice
					JOIN negozio n ON l.negozio = n.codice
			WHERE 	p.codice NOT IN (SELECT idProdotto FROM prodotti_piu_citta)
			AND p.codice=%s;''', (idProdotto, ))
		results = list(cur)

		# se la lista e' vuota, il prodotto esiste ma non e' una esclusiva, oppure
		# non e' in vendita presso nessun negozio
		if !results:
			print('Il prodotto non e\' una esclusiva oppure non e\' in vendita')

		else:
			print('il prodotto e\' in vendita esclusiva presso', cur.fetchone()[0]);

File: 3° Anno/Basi Di Dati/Laboratorio/Michael_Posenato_Esami_Svolti/• 2019 19 Giugno/es3.java
/*	programma java che leggendo il codice di una sezione da console, visualizzi
	il risultato della query 2a. Se il codice non esiste, il programma deve
	richiedere nuovamente il dato	*/

public static void main(String[] args) throws SQLException, ClassNotFoundException {

	Scanner 	in = new Scanner(System.in);
	String 		sezione = '';
	ResultSet 	rs = null;
	preparedStatement pst = null;

	try(Class.forName("org.posrgresql.Driver");
		Connection conn = DriverManager.getConnection("jdbc:postgresql://server_name:port/database", user, password)){

		// richiesta e controllo sull'user input
		do {
			System.out.print("Codice sezione: ");
			sezione = in.nextLine();

			try (pst = conn.prepareStatement("SELECT COUNT(*) FROM sezione WHERE codice=?")) {
				pst.setString(1, sezione);

				rs = pst.executeQuery();
				rs.next();

				if (rs.getInt(1) != 1) {
					System.out.print("Codice sezione inserito non valido, riprovare.");
					continue;
				}
			}catch(SQLException e){
				System.out.print("Errore nel controllo input utente: " e.getMessage());
			}

		}while(false)

		// codice sezione valido, eseguo la query
		try (pst = conn.executeQuery("<CODICE QUERY ES2A> AND s.codice=?")) {
			pst.setString(1, sezione);
			rs = pst.executeQuery();

			// display del risultato
			while(rs.next()) {
				System.out.println(String.format("%20s %20s %20s %20s %d", rs.getString(1), rs.getString(2), rs.getString(3), rs.getString(4), rs.getInt(5)));
			}
		}catch(SQLException e) {
			System.out.print("Errore nel fetch del risultato: " + e.getMessage());
		}

	}catch(SQLException e){
		System.out.print("Errore nella creazione della connessione: " + e.getMessage());
	}catch(ClassNotFoundException e){
		System.out.print("Errore nel fetch dei driver: " + e.getMessage());
	}
}